drivers/pwm/pwm_mspm0.c

```diff
@@ -0,0 +1,137 @@
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(pwm_mspm0, CONFIG_LOG_DEFAULT_LEVEL);
+
+#define DT_DRV_COMPAT ti_mspm0_pwm
+
+struct pwm_mspm0_config {
+	GPTIMER_Regs *timer;
+	const struct pinctrl_dev_config *pinctrl;
+	const uint32_t frequency;
+	const DL_Timer_ClockConfig clock_cfg;
+	const DL_TIMER_PWM_MODE pwm_mode;
+	const uint8_t cc_index;
+	const uint8_t direction;
+	const bool is_tima;
+	const bool four_cc;
+};
+
+static int pwm_mspm0_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					uint64_t *cycles)
+{
+	const struct pwm_mspm0_config *config = dev->config;
+	*cycles = config->frequency /
+		  ((config->clock_cfg.divideRatio + 1) * (config->clock_cfg.prescale + 1));
+	return 0;
+}
+
+static int pwm_mspm0_set_cycles(const struct device *dev, uint32_t channel, uint32_t period_cycles,
+				uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	const struct pwm_mspm0_config *cfg = dev->config;
+	if (pulse_cycles > 100) {
+		LOG_DBG("Invalid duty cycles, range should be [0, 100]");
+		return -1;
+	}
+
+	// driver specific to stop the counter
+	if (flags == 1) {
+		DL_Timer_stopCounter(cfg->timer);
+		return 0;
+	}
+
+	DL_Timer_startCounter(cfg->timer);
+	DL_Timer_setLoadValue(cfg->timer, period_cycles);
+	DL_Timer_setCaptureCompareValue(cfg->timer, (pulse_cycles * period_cycles) / 100,
+					cfg->cc_index);
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_mspm0_driver_api = {
+	.set_cycles = pwm_mspm0_set_cycles,
+	.get_cycles_per_sec = pwm_mspm0_get_cycles_per_sec,
+};
+
+static int pwm_mspm0_init(const struct device *dev)
+{
+	const struct pwm_mspm0_config *cfg = dev->config;
+	int ret = pinctrl_apply_state(cfg->pinctrl, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		LOG_ERR("Error, failed to apply pinctrl: %d", ret);
+		return ret;
+	}
+
+	DL_Timer_reset(cfg->timer);
+	DL_Timer_enablePower(cfg->timer);
+	DL_Timer_setClockConfig(cfg->timer, (DL_Timer_ClockConfig *)&cfg->clock_cfg);
+
+	// TIMG/TIMA have different struct and init functions. Also we do not need to store
+	// the configuration for each module
+	if (cfg->is_tima) {
+		// the period value doesn't matter, but it is a req to setup correctly the pwm
+		// we will eventually set the correct period by calling set_cycles
+		const DL_TimerA_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.startTimer = DL_TIMER_STOP,
+			.isTimerWithFourCC = cfg->four_cc,
+			.period = 1000,
+		};
+
+		DL_TimerA_initPWMMode(cfg->timer, (DL_TimerA_PWMConfig *)&pwm_cfg);
+	} else {
+		const DL_TimerG_PWMConfig pwm_cfg = {
+			.pwmMode = cfg->pwm_mode,
+			.startTimer = DL_TIMER_STOP,
+		};
+
+		DL_TimerG_initPWMMode(cfg->timer, (DL_TimerG_PWMConfig *)&pwm_cfg);
+	}
+
+	DL_Timer_setCaptureCompareOutCtl(cfg->timer, DL_TIMER_CC_OCTL_INIT_VAL_HIGH,
+					 DL_TIMER_CC_OCTL_INV_OUT_ENABLED,
+					 DL_TIMER_CC_OCTL_SRC_FUNCVAL, cfg->cc_index);
+
+	DL_Timer_setCaptCompUpdateMethod(cfg->timer, DL_TIMER_CC_UPDATE_METHOD_IMMEDIATE,
+					 cfg->cc_index);
```

Comment:
I think the generic init function sets these up sensibly (or they have adequate defaults.
Unless there is a good reason to configure them like this, I think you should strip it out.

Reply:
you are right